Add cooperative groups (#3478)
ROCm-Developer-Tools/HIP: docs/how-to/cooperative_groups.rst

.. meta::
  :description: This chapter describe how to use cooperative groups in HIP
  :keywords: AMD, ROCm, HIP, cooperative groups

*******************************************************************************
Cooperative Groups in HIP
*******************************************************************************

> The Cooperative Groups is an extension of the exsiting ROCm programming model, 

Reviewer: ```suggestion
The Cooperative Groups is an extension of the existing ROCm programming model, 
```

to get a more flexible grouping mechanism for the Developers. This feature was 
introduced with CUDA 9 at NVIDIA GPUs and with ROCm 4.1 at AMD GPUs.

> The API accessable in the ``cooperative_groups`` namespace after the 

Reviewer: ```suggestion
The API accessible in the ``cooperative_groups`` namespace after the 
```

``cooperative_groups.h`` is included. The header contains the following 
elements:

* Data types for representing groups
* Operations to generate implicit groups defined;
* Collectives for partitioning existing groups into new groups;
* Operation to synchronize all threads within the group;
* Operations to inspect the group properties;
* Collectives that expose low-level, group-specific and often HW accelerated, operations.

> The code differenc to the original block model can be find in the following table. 

Reviewer: ```suggestion
The code difference to the original block model can be found in the following table. 
```
typo

              if(group_thread_id < i)
                val += shared[group_thread_id + i];

              // Synchronize all threads in the group
              g.sync();
            }

            // ...
          }

    * - .. code-block:: C++

          __global__ void sum_kernel(...) {
            // ...
    
            // Workspace array in shared memory
            __shared__ unsigned int workspace[2048];

            // ...

            // Perform reduction
            output = reduce_sum(workspace, input);

            // ...
          }

      - .. code-block:: C++

          __global__ void sum_kernel(...) {
            // ...

            // Workspace array in shared memory
            __shared__ unsigned int workspace[2048];

            // ...

            thread_block thread_block_group = this_thread_block();
            // Perform reduction
            output = reduce_sum(thread_block_group, workspace, input);

            // ...
          }

The kernel launch also different at cooperative groups case, which depends on the 
group type. For example, at grid groups with single GPU case the ``hipLaunchCooperativeKernel``
has to be used.

Group Types
=============

> There are different group types for based on different level on grouping.

Reviewer: ```suggestion
There are different group types based on different levels of grouping.
```

.. code-block:: C++
  
  class thread_block;

Constructed via:

.. code-block:: C++
  
  thread_block g = this_thread_block();

The ``group_index()`` , ``thread_index()`` , ``thread_rank()`` , ``size()``, ``cg_type()``,  
``is_valid()`` , ``sync()`` and ``group_dim()`` member functions are public of the 
thread_block class. For further details check the :ref:`thread_block references <thread_block_ref>` . 

Grid Group
------------

Represents an inter-workgroup cooperative group type where the participating threads
within the group spans across multiple workgroups running the (same) kernel on the same device.
To be able to synchronize across the grid, you need to use the cooperative launch API.

.. code-block:: C++

  class grid_group;

Constructed via:

.. code-block:: C++

  grid_group g = this_grid();

The ``thread_rank()`` , ``size()``, ``cg_type()``, ``is_valid()`` and ``sync()`` member functions
are public of the ``grid_group`` class. For further details check the :ref:`grid_group references <grid_group_ref>`. 

Multi Grid Group
------------------

Represents an inter-device cooperative group type where the participating threads

> within the group spans across multple devices, running the (same) kernel on these devices

Reviewer: ```suggestion
within the group spans across multiple devices, running the (same) kernel on these devices
```

All the mutli grid group APIs require that you have used the appropriate launch API.

.. code-block:: C++

  class multi_grid_group;

Constructed via:

.. code-block:: C++

  // Kernel must be launched with the cooperative multi-device API
  multi_grid_group g = this_multi_grid();

The ``num_grids()`` , ``grid_rank()`` , ``thread_rank()``, ``size()``, ``cg_type()``, ``is_valid()`` ,
and ``sync()`` member functions are public of the ``multi_grid_group`` class. For
further details check the :ref:`multi_grid_group references <multi_grid_group_ref>` . 

Thread Block Tile
------------------

> This constructs a templated class derieved from thread_group. The template defines tile

Reviewer: ```suggestion
This constructs a templated class derived from ``thread_group``. The template defines tile
```

size of the new thread group at compile time.

.. code-block:: C++

  template <unsigned int Size, typename ParentT = void>
  class thread_block_tile;

Constructed via:

.. code-block:: C++
  
  template <unsigned int Size, typename ParentT>
  _CG_QUALIFIER thread_block_tile<Size, ParentT> tiled_partition(const ParentT& g)

.. note::
  
  * ``Size`` must be a power of 2 and not bigger than wavefront size.
  * ``shfl`` functions are currently supports integer or float type.

The ``thread_rank()`` , ``size()``, ``cg_type()``, ``is_valid()``, ``sync()``, 
``meta_group_rank()``, ``meta_group_size()``, ``shfl(...)``, ``shfl_down(...)``, 
``shfl_up(...)`` and ``shfl_xor(...)`` member functions are public of the ``thread_block_tile``
class. For further details check the :ref:`thread_block_tile references <thread_block_tile_ref>` . 

Coalesced Groups
------------------

> Represents a active thread group in a wavefront. This group type also supports sub-wave level

Reviewer: ```suggestion
Represents an active thread group in a wavefront. This group type also supports sub-wave level
```

intrinsics.

.. code-block:: C++

  class coalesced_group;

Constructed via:

.. code-block:: C++

  coalesced_group active = coalesced_threads();

.. note::
  
  * ``shfl`` functions are currently supports integer or float type.

The ``thread_rank()`` , ``size()``, ``cg_type()``, ``is_valid()``, ``sync()``, 
``meta_group_rank()``, ``meta_group_size()``, ``shfl(...)``, ``shfl_down(...)``, 
and ``shfl_up(...)`` member functions are public of the ``coalesced_group`` class. 
For further details check the :ref:`coalesced_group references <coalesced_group_ref>` .

Synchronization
=================

> At different type of gourps the synchronization requires to used the correct cooperative group

Reviewer: ```suggestion
At different types of groups, the synchronization requires using the correct cooperative group
```

.. meta::
  :description: This topic describes how to use cooperative groups in HIP
  :keywords: AMD, ROCm, HIP, cooperative groups

*******************************************************************************
Cooperative Groups in HIP
*******************************************************************************

Cooperative Groups is an extension to the ROCm programming model. It provides developers with a flexible grouping mechanism. This feature was introduced with AMD ROCm 4.1 and NVIDIA CUDA.

The API is accessible in the ``cooperative_groups`` namespace after the 
``cooperative_groups.h`` is included. The header contains the following 
elements:

* Data types for representing groups
* Operations to generate implicit groups defined
* Collectives for partitioning existing groups into new groups
* Operation to synchronize all threads within the group
* Operations to inspect the group properties
* Collectives that expose low-level, group-specific, and, often, hardware-accelerated operations.

> You can find the code difference to the original block model in the following table. 

Reviewer 2: ```suggestion
You'll find the code difference to the original block model in the following table. 
```

        launchParamsList[deviceID].sharedMem = 0;
        launchParamsList[deviceID].stream = stream;
        launchParamsList[deviceID].args = params;
    }

    HIP_CHECK(hipLaunchCooperativeKernelMultiDevice(launchParamsList,
                                                    (int)deviceIDs.size(),
                                                    hipCooperativeLaunchMultiDeviceNoPreSync));

3. The device side code of the grid synchronization over multiple GPUs:

.. code-block:: C++

  multi_grid_group multi_grid = this_multi_grid();
  multi_grid.sync();

Unsupported NVIDIA CUDA features
===========================

HIP doesn't support the following NVIDIA CUDA optional headers:

.. code-block:: C++

    // Optionally include for memcpy_async() collective
    #include <cooperative_groups/memcpy_async.h>
    // Optionally include for reduce() collective
    #include <cooperative_groups/reduce.h>
    // Optionally include for inclusive_scan() and exclusive_scan() collectives
    #include <cooperative_groups/scan.h>

HIP doesn't support the following CUDA class the in ``cooperative_groups`` namespace:

* ``cluster_group``

> 

Reviewer 2: Something is messed up here. 
Perhaps the second highlighted line should be deleted. 
![image](https://github.com/ROCm/HIP/assets/164940351/a5b29d82-9643-45db-a4ec-3617d9960b2d)

Author: Fixed.

.. meta::

>   :description: This chapter describe how to use cooperative groups in HIP

Reviewer 2: ```suggestion
  :description: This topic describes how to use cooperative groups in HIP
```

  :keywords: AMD, ROCm, HIP, cooperative groups

*******************************************************************************
Cooperative Groups in HIP
*******************************************************************************

> The Cooperative Groups is an extension of the existing ROCm programming model, 
> to get a more flexible grouping mechanism for the Developers. This feature was 
> introduced with CUDA 9 at NVIDIA GPUs and with ROCm 4.1 at AMD GPUs.

Reviewer 2: ```suggestion
Cooperative Groups is an extension to the ROCm programming model. It provides developers with a flexible grouping mechanism. This feature was introduced with AMD ROCm 4.1 and NVIDIA CUDA.
```

> The API accessible in the ``cooperative_groups`` namespace after the 

Reviewer 2: ```suggestion
The API is accessible in the ``cooperative_groups`` namespace after the 
```

> ``cooperative_groups.h`` is included. The header contains the following 
> elements:
> 
> * Data types for representing groups
> * Operations to generate implicit groups defined;
> * Collectives for partitioning existing groups into new groups;
> * Operation to synchronize all threads within the group;
> * Operations to inspect the group properties;
> * Collectives that expose low-level, group-specific and often hardware accelerated, operations.
> 
> The code difference to the original block model can be found in the following table. 

Reviewer 2: ```suggestion
``cooperative_groups.h`` is included. The header contains the following 
elements:

* Data types for representing groups
* Operations to generate implicit groups defined
* Collectives for partitioning existing groups into new groups
* Operation to synchronize all threads within the group
* Operations to inspect the group properties
* Collectives that expose low-level, group-specific, and, often, hardware-accelerated operations.

You can find the code difference to the original block model in the following table. 
```

.. list-table:: Cooperative Group Example
    :header-rows: 1
    :widths: 50,50

    * - **Original Block**
      - **Cooperative Groups**

    * - .. code-block:: C++
          
          __device__ int reduce_sum(int *shared, int val) {
              
              // Thread ID
              const unsigned int thread_id = threadIdx.x;

              // Every iteration the number of active threads halves,
              // until we processed all values
              for(unsigned int i = blockDim.x / 2; i > 0; i /= 2) {

>                 // Store value in shared memroy with thread ID

Reviewer 2: ```suggestion
                // Store value in shared memory with thread ID
```

              // Store value in shared memroy with thread ID
              shared[group_thread_id] = val;

              // Synchronize all threads in the group
              g.sync();

              // Active thread sum up
              if(group_thread_id < i)
                val += shared[group_thread_id + i];

              // Synchronize all threads in the group
              g.sync();
            }

            // ...
          }

    * - .. code-block:: C++

          __global__ void sum_kernel(...) {
            // ...
    
            // Workspace array in shared memory
            __shared__ unsigned int workspace[2048];

            // ...

            // Perform reduction
            output = reduce_sum(workspace, input);

            // ...
          }

      - .. code-block:: C++

          __global__ void sum_kernel(...) {
            // ...

            // Workspace array in shared memory
            __shared__ unsigned int workspace[2048];

            // ...

            thread_block thread_block_group = this_thread_block();
            // Perform reduction
            output = reduce_sum(thread_block_group, workspace, input);

            // ...
          }

> The kernel launch also different at cooperative groups case, which depends on the 
> group type. For example, at grid groups with single GPU case the ``hipLaunchCooperativeKernel``
> has to be used.

Reviewer 2: ```suggestion
The kernel launch also differs with cooperative groups as it depends on the group type. For example, grid groups with a single-GPU, the ``hipLaunchCooperativeKernel`` has to be used.
```

Group Types
=============

> There are different group types based on different level of synchronization and
> data sharing among threads. 

Reviewer 2: ```suggestion
Group types are based on the levels of synchronization and data sharing among threads. 
```

> Thread Block Group
> --------------------
> 
> Represents an intra-workgroup cooperative group type where the
> participating threads within the group are exactly the same threads
> which are participated in the currently executing ``workgroup``.

Reviewer 2: ```suggestion
Thread-block group
--------------------

Represents an intra-workgroup cooperative group type where the participating threads within the group are the same threads that participated in the currently executing ``workgroup``.
```

.. code-block:: C++
  
  class thread_block;

Constructed via:

.. code-block:: C++
  
  thread_block g = this_thread_block();

The ``group_index()`` , ``thread_index()`` , ``thread_rank()`` , ``size()``, ``cg_type()``,  
``is_valid()`` , ``sync()`` and ``group_dim()`` member functions are public of the 
thread_block class. For further details check the :ref:`thread_block references <thread_block_ref>` . 

> Grid Group
> ------------
> 
> Represents an inter-workgroup cooperative group type where the participating threads
> within the group spans across multiple workgroups running the (same) kernel on the same device.
> To be able to synchronize across the grid, you need to use the cooperative launch API.

Reviewer 2: ```suggestion
Grid group
------------

Represents an inter-workgroup cooperative group type where the group's participating threads span multiple workgroups running the same kernel on the same device. Use the cooperative launch API to synchronize across the grid.
```

.. code-block:: C++

  class grid_group;

Constructed via:

.. code-block:: C++

  grid_group g = this_grid();

The ``thread_rank()`` , ``size()``, ``cg_type()``, ``is_valid()`` and ``sync()`` member functions
are public of the ``grid_group`` class. For further details check the :ref:`grid_group references <grid_group_ref>`. 

> Multi Grid Group
> ------------------
> 
> Represents an inter-device cooperative group type where the participating threads
> within the group spans across multiple devices, running the (same) kernel on these devices
> All the multi grid group APIs require that you have used the appropriate launch API.

Reviewer 2: ```suggestion
Multi-grid group
------------------

Represents an inter-device cooperative group type where the participating threads within the group span multiple devices that run the same kernel on the devices. All the multi-grid group APIs require that you have used the appropriate launch API.
```

.. code-block:: C++

  class multi_grid_group;

Constructed via:

.. code-block:: C++

  // Kernel must be launched with the cooperative multi-device API
  multi_grid_group g = this_multi_grid();

The ``num_grids()`` , ``grid_rank()`` , ``thread_rank()``, ``size()``, ``cg_type()``, ``is_valid()`` ,
and ``sync()`` member functions are public of the ``multi_grid_group`` class. For
further details check the :ref:`multi_grid_group references <multi_grid_group_ref>` . 

> Thread Block Tile
> ------------------
> 
> This constructs a templated class derived from ``thread_group``. The template defines tile
> size of the new thread group at compile time.

Reviewer 2: ```suggestion
Thread-block tile
------------------

This constructs a templated class derived from ``thread_group``. The template defines the tile
size of the new thread group at compile time.
```

.. code-block:: C++

  template <unsigned int Size, typename ParentT = void>
  class thread_block_tile;

Constructed via:

.. code-block:: C++
  
  template <unsigned int Size, typename ParentT>
  _CG_QUALIFIER thread_block_tile<Size, ParentT> tiled_partition(const ParentT& g)

.. note::
  

>   * ``Size`` must be a power of 2 and not bigger than wavefront size.
>   * ``shfl`` functions are currently supports integer or float type.

Reviewer 2: ```suggestion
  * ``Size`` must be a power of 2 and not larger than wavefront size.
  * ``shfl`` functions support integer or float type.
```

The ``thread_rank()`` , ``size()``, ``cg_type()``, ``is_valid()``, ``sync()``, 
``meta_group_rank()``, ``meta_group_size()``, ``shfl(...)``, ``shfl_down(...)``, 
``shfl_up(...)`` and ``shfl_xor(...)`` member functions are public of the ``thread_block_tile``
class. For further details check the :ref:`thread_block_tile references <thread_block_tile_ref>` . 

> Coalesced Groups

Reviewer 2: ```suggestion
Coalesced groups
```

------------------

Represents an active thread group in a wavefront. This group type also supports sub-wave level
intrinsics.

.. code-block:: C++

  class coalesced_group;

Constructed via:

.. code-block:: C++

  coalesced_group active = coalesced_threads();

.. note::
  

>   * ``shfl`` functions are currently supports integer or float type.

Reviewer 2: ```suggestion
  * ``shfl`` functions support integer or float type.
```

The ``thread_rank()`` , ``size()``, ``cg_type()``, ``is_valid()``, ``sync()``, 
``meta_group_rank()``, ``meta_group_size()``, ``shfl(...)``, ``shfl_down(...)``, 
and ``shfl_up(...)`` member functions are public of the ``coalesced_group`` class. 

> For further details check the :ref:`coalesced_group references <coalesced_group_ref>` .

Reviewer 2: ```suggestion
For more information, see :ref:`coalesced_group references <coalesced_group_ref>` .
```

Synchronization
=================

> At different types of groups, the synchronization requires using the correct cooperative group

Reviewer 2: ```suggestion
With each group type, the synchronization requires using the correct cooperative group
```

launch API.

> Thread-Block Synchronization
> -----------------------------------------------
> 
> 1. The new block representation can be accessed with the original kernel launch methods.
> 2. The device side synchronization is written in the following form.
> 

Reviewer 2: ```suggestion
Thread-block synchronization
-----------------------------------------------

1. You can access the new block representation using the original kernel launch methods.
2. The device side synchronization is written in the following form:

```

.. code-block:: C++

  thread_block g = this_thread_block();
  g.sync();

> Grid Synchronization
> ---------------------
> 
> This section describes the necessary step to be able to synchronize group over a single GPU:
> 
> 1. Check the cooperative launch capability on single AMD GPU:

Reviewer 2: ```suggestion
Grid Synchronization
---------------------

This section describes the steps to synchronize the group over a single GPU:

1. Confirm the cooperative launch capability on the single AMD GPU:
```

.. code-block:: C++

    int device               = 0;
    int supports_coop_launch = 0;
    // Check support
    // Use hipDeviceAttributeCooperativeMultiDeviceLaunch when launching across multiple devices
    HIP_CHECK(hipGetDevice(&device));
    HIP_CHECK(
        hipDeviceGetAttribute(&supports_coop_launch, hipDeviceAttributeCooperativeLaunch, device));
    if(!supports_coop_launch)
    {
        std::cout << "Skipping, device " << device << " does not support cooperative groups"
                  << std::endl;
        return 0;
    }

> 2. Launch the cooperative kernel on single GPU:

Reviewer 2: ```suggestion
2. Launch the cooperative kernel on a single GPU:
```

.. code-block:: C++

    void* params[] = {&d_vector, &d_block_reduced, &d_partition_reduced};
    // Launching kernel from host.
    HIP_CHECK(hipLaunchCooperativeKernel(vector_reduce_kernel<partition_size>,
                                         dim3(num_blocks),
                                         dim3(threads_per_block),
                                         params,
                                         0,
                                         hipStreamDefault));

> 3. The device side synchronization on single GPU:

Reviewer 2: Don't understand this - what is the procedure?

Author: This is actual synchronization of the threads over the grid. This section describes the necessary steps to be able to synchronize with the different types of groups.

.. code-block:: C++

  grid_group grid = this_grid();
  grid.sync();

> Multi-Grid Synchronization

Reviewer 2: ```suggestion
Multi-grid Synchronization
```

-----------------------------

> This section describes the necessary step to be able to synchronize group over multiple GPUs:

Reviewer 2: ```suggestion
This section describes how to synchronize the group over multiple GPUs:
```

> 1. Check the cooperative launch capability over multiple GPUs:

Reviewer 2: ```suggestion
1. Confirm the cooperative launch capability over multiple GPUs:
```

  }

2. Launch the cooperative kernel over multiple GPUs:

.. code-block:: C++

    hipLaunchParams *launchParamsList = (hipLaunchParams*)malloc(sizeof(hipLaunchParams) * deviceIDs.size());
    for(int deviceID : deviceIDs) {
		// Set device
		HIP_CHECK(hipSetDevice(deviceID));
		
		// Create stream
        hipStream_t stream;
        HIP_CHECK(hipStreamCreate(&stream));
		
		// Parameters
		void* params[] = {&(d_vector[deviceID]), &(d_block_reduced[deviceID]), &(d_partition_reduced[deviceID])};
		
		// Set launchParams
        launchParamsList[deviceID].func = (void*)vector_reduce_kernel<partition_size>;
        launchParamsList[deviceID].gridDim = dim3(1);
        launchParamsList[deviceID].blockDim = dim3(threads_per_block);
        launchParamsList[deviceID].sharedMem = 0;
        launchParamsList[deviceID].stream = stream;
        launchParamsList[deviceID].args = params;
    }

    HIP_CHECK(hipLaunchCooperativeKernelMultiDevice(launchParamsList,
                                                    (int)deviceIDs.size(),
                                                    hipCooperativeLaunchMultiDeviceNoPreSync));

> 3. The device side synchronization over multiple GPUs:

Reviewer 2: Is this a procedure - what is this supposed to mean?

Author: This is actual synchronization of the threads over the grid. This section describes the necessary steps to be able to synchronize with the different types of groups.

.. code-block:: C++

  multi_grid_group multi_grid = this_multi_grid();
  multi_grid.sync();

> Unsupported CUDA features
> ===========================
> 
> The following CUDA optional headers are not supported on HIP:

Reviewer 2: ```suggestion
Unsupported NVIDIA CUDA features
===========================

HIP doesn't support the following NVIDIA CUDA optional headers:
```

.. code-block:: C++

    // Optionally include for memcpy_async() collective
    #include <cooperative_groups/memcpy_async.h>
    // Optionally include for reduce() collective
    #include <cooperative_groups/reduce.h>
    // Optionally include for inclusive_scan() and exclusive_scan() collectives
    #include <cooperative_groups/scan.h>

> The following CUDA classes in ``cooperative_groups`` namespace are not supported on HIP:

Reviewer 2: ```suggestion
HIP doesn't support the following CUDA class the in ``cooperative_groups`` namespace:
```

* ``cluster_group``

> The following CUDA functions/operators in ``cooperative_groups`` namespace are not supported on HIP:

Reviewer 2: ```suggestion
HIP doesn't support the following CUDA functions/operators in ``cooperative_groups`` namespace:

HIP doesn't support the following
```

.. meta::
  :description: This topic describes how to use cooperative groups in HIP
  :keywords: AMD, ROCm, HIP, cooperative groups

*******************************************************************************
Cooperative Groups
*******************************************************************************

> Cooperative groups API is an extension to the ROCm programming model. It provides developers with a flexible grouping mechanism. This feature was introduced with ROCm 4.1 at AMD platform and CUDA 9.0 at NVIDIA platform.

Reviewer 2: Do we need to discuss when this feature was added. Seems unnecessary.

Author: Removed.

The API is accessible in the ``cooperative_groups`` namespace after the  ``hip_cooperative_groups.h`` is included. The header contains the following  elements:

* Static functions to create groups and subgroups.
* Hardware-accelerated operations over the whole group, like shuffles.
* Data types of cooperative groups.
* Synchronize member function of the groups.
* Get group properties member functions.

Cooperative groups thread model
===============================

The thread hierarchy abstraction of cooperative groups is in
:numref:`coop_thread_hierarchy`.

.. _coop_thread_hierarchy:

.. figure:: ../data/how-to/cooperative_groups/thread_hierarchy_coop.svg
  :alt: Diagram depicting nested rectangles of varying color. The outermost one
        titled "Grid", inside sets of different sized rectangles layered on
        one another titled "Block". Each "Block" containing sets of uniform
        rectangles layered on one another titled "Warp". Each of the "Warp"
        titled rectangles filled with downward pointing arrows inside.

>   Cooperative group thread hierarchy.

Reviewer 3: This isn't really the entire hierarchy.  Maybe call it "top of cooperative groups thread hierarchy"?

further details check the :ref:`multi_grid_group references <multi_grid_group_ref>` .

Thread-block tile
------------------

This constructs a templated class derived from ``thread_group``. The template defines the tile
size of the new thread group at compile time. This group type also supports sub-wave level intrinsics.

.. code-block:: cpp

  template <unsigned int Size, typename ParentT = void>
  class thread_block_tile;

Constructed via:

.. code-block:: cpp

  template <unsigned int Size, typename ParentT>
  _CG_QUALIFIER thread_block_tile<Size, ParentT> tiled_partition(const ParentT& g)

.. note::

  * Size must be a power of 2 and not larger than warp (wavefront) size.
  * ``shfl()`` functions support integer or float type.

The ``thread_rank()`` , ``size()``, ``cg_type()``, ``is_valid()``, ``sync()``, ``meta_group_rank()``, ``meta_group_size()``, ``shfl()``, ``shfl_down()``, ``shfl_up()``, ``shfl_xor()``, ``ballot()``, ``any()``, ``all()``, ``match_any()`` and ``match_all()`` member functions are public of the ``thread_block_tile`` class. For further details, check the :ref:`thread_block_tile references <thread_block_tile_ref>` .

Coalesced groups
------------------

> Threads (64 threads on CDNA and 32 threads on RDNA or NVIDIA GPUs) in a warp cannot execute different instructions simultaneously, so conditional branches are executed serially within the warp. When threads encounter a conditional branch, they can diverge, resulting in some threads being disabled, if they do not meet the condition to execute that branch. The active threads referred as coalesced, and coalesced group represents an active thread group within a warp. 

Reviewer 3: I wouldn't mention nvidia here and will note that nvidia independent thread scheduling presents the appearance that threads on different branches execute concurrently.  We need a warning here that we do not support independent thread scheduling and some cuda apps using them ported to hip can deadock if they make use of independent thread scheduling.

.. meta::
  :description: This topic describes how to use cooperative groups in HIP
  :keywords: AMD, ROCm, HIP, cooperative groups

.. _cooperative_groups_how-to:

*******************************************************************************
Cooperative Groups
*******************************************************************************

> Cooperative groups API is an extension to the HIP programming model, which provides developers with a flexible, dynamic grouping mechanism for the communicating threads. The API enables the developers to specify the level of granularity for thread communication which can lead to more efficient parallel decompositions.

Reviewer 4: "for the communicating threads"
 
 What are the communicating threads? Are they a special type of thread Cooperative Groups API applies to?

Author: No. The cooperative groups API focusing on the threads trying to exchange information (communicating).

> .. Maybe this sentence is better: The rich set of APIs introduced by Cooperative Groups allow the programmer to define their own set of thread groups which may fit their user-cases better than those defined by the hardware.

Reviewer 4: This is good.

Author: We solved.

The API is accessible in the ``cooperative_groups`` namespace after the  ``hip_cooperative_groups.h`` is included. The header contains the following  elements:

> * Data types of cooperative groups.

Reviewer 4: * Static functions to create groups and subgroups.
* Hardware-accelerated operations over the whole group, like shuffles.
* Data types of cooperative groups.
* Synchronize member function of the groups.
* Get group properties member functions.

Author: Fixed.

* Synchronize member function of the groups.
* Get group properties member functions.
* Hardware-accelerated operations over the whole group, like shuffles.
* Static functions to create groups and subgroups.

.. note::

  This feature was introduced with ROCm 4.1 at AMD platform and CUDA 9.0 at NVIDIA platform. This introduced a new level between block and thread block level for synchronization.

Cooperative groups thread model
===============================

The thread hierarchy abstraction of cooperative groups is in
:numref:`coop_thread_hierarchy`.

.. _coop_thread_hierarchy:

.. figure:: ../data/how-to/cooperative_groups/thread_hierarchy_coop.svg
  :alt: Diagram depicting nested rectangles of varying color. The outermost one
        titled "Grid", inside sets of different sized rectangles layered on
        one another titled "Block". Each "Block" containing sets of uniform
        rectangles layered on one another titled "Warp". Each of the "Warp"
        titled rectangles filled with downward pointing arrows inside.

  Cooperative group thread hierarchy.

> The **multi grid** is an abstraction of potentially multiple simultaneous launches of the same kernel over multiple devices (Deprecated since 5.0). The **grid** in cooperative groups is a single dispatch of kernels for execution like the original grid. The ability to synchronize over a grid requires the kernel to be launched using the cooperative groups API. The **block** is the same as the :ref:`inherent_thread_model` block entity.

Reviewer 4: I don't understand the need to document a deprecated feature. Especially one that has been deprecated since 5.0? 

"The ability to synchronize a grid requires the kernel to be launched using the Cooperative Groups API." This would seem to be a significant feature of Cooperative Groups. maybe we should lead with this?

Author: Moved outside into a note.

.. note::

  Explicit warp-level thread handling is absent from the Cooperative Groups API.
  In order to exploit the known hardware SIMD width on which built-in

>   functionality translates to simpler logic, one may use the group partitioning

Reviewer 4: functionality translates to simpler logic, you can use the group partitioning

Author: Fixed.

  part of the API, such as ``tiled_partition``.

For details on memory model, check the :ref:`memory model description <memory_hierarchy>`.

> Cooperative groups simple example

Reviewer 4: I would move this "simple example" to after the discussion of the features and functions of Cooperative Groups. The reason being that the features and functions have not been explained yet, and will be unfamiliar to the user reviewing the code examples.

Author: Moved lower.

          thread_block thread_block_group = this_thread_block();
          // Perform reduction
          output = reduce_sum(thread_block_group, workspace, input);

          // ...
      }

At the device function, the input group type is the ``thread_group``, which is the parent class of all the cooperative groups type. With this, you can write generic functions, which can work with any type of cooperative groups.

Group types
===========

Group types are based on the levels of synchronization and data sharing among threads.

Thread-block group
------------------

Represents an intra-block cooperative groups type where the participating threads within the group are the same threads that participated in the currently executing ``block``.

.. code-block:: cpp

  class thread_block;

Constructed via:

.. code-block:: cpp

  thread_block g = this_thread_block();

The ``group_index()`` , ``thread_index()`` , ``thread_rank()`` , ``size()``, ``cg_type()``, ``is_valid()`` , ``sync()`` and ``group_dim()`` member functions are public of the thread_block class. For further details, check the :ref:`thread_block references <thread_block_ref>` .

Grid group
------------

> Represents an inter-block cooperative groups type where the group's participating threads span multiple blocks running the same kernel on the same device. Use the cooperative launch API to synchronize across the grid.

Reviewer 4: Use the cooperative launch API to synchronize across the grid.

Should this be: Use the cooperative launch API to enable synchronization across the grid group?

Author: Fixed.
